# Variables and Data Types in Python

## What are variables in Python?
In Python, when an object is assigned to a variable, the value of that object is stored in that variable.

To create a variable in Python, type

```python3
{variable_name} = {object}
```

For example, to assign the object/value 30 to the variable n, type

In [3]:
n = 30

The value 30 is now stored in the variable n. To confirm this, we can type

In [4]:
print(n)

30


Since the value 30 is assigned to the variable n, printing n outputs the value 30. 

## Why do we need variables?
Variables are useful tools as they allow the user to create programs that can be used with a variety of datasets. Rather than creating a program with values assigned to a single dataset, variables can be utilized so that users can enter different data values into a single program. This allows for greater flexibility in programming. 

## What are data types?
When an object is assigned to a variable, the variable is given a specific *data type* that corresponds to the object. To view the data type of a variable, type

```python3 
type({variable_name})
```

4 basic data types are integer (int), float, string (str), and boolean (bool). 

### Integer (Int)
The *int* data type refers to numerical objects that are integers. For example, we earlier defined the variable n as 30. Since 30 is an integer, the data type of n is *int*. We can confirm this by typing

In [5]:
type(n)

int

As we expected, the output is int. 

### Float
The *float* data type is assigned to numerical objects that have a decimal point. For example, let's define a new variable *var*. 

In [8]:
var = 3.14

Since 3.14 has a decimal place, we should expect the data type of var to be float. We can check this by typing

In [7]:
type(var)

float

As expected, checking the data type of *var* returned the output *float*. 

### String (Str)
A *string* in Python refers to objects that consist of a sequence of characters, digits, and/or symbols. Strings are denoted by single quotes (i.e. '{string}'). For example, let's define a new variable *my_var*. 

In [9]:
my_var = 'This variable is a string.'

Since the object of my_var is a string of characters, we should expect the data type of my_var to be string (str). We can check this by typing

In [10]:
type(my_var)

str

As expected, the output is str. 

### Boolean
The boolean (bool) data type refers to two conditional values, *True* and *False*. These values are often used when testing variables within a program.

To show that True and False are boolean data types, we can type

In [20]:
type(True)

bool

In [21]:
type(False)

bool

**NOTE** that True and False are not entered with quotes. This would change their data type to str. 

## Why do data types exist?
Data types give the operating system information on how to interpret the value of an object. For example, when the object 3.14 is assigned to a variable, it is interpreted as a numerical value (specifically a float). However, when the object '3.14' is assigned to a variable, it is interpreted as a string of text. To illustrate this concept, let's define a new variable new_var

In [14]:
new_var = '3.14'

When we add 2.0 to our previously defined float variable *var*, the two numerical values are added together. 

In [16]:
var + 2.0

5.140000000000001

However, when we try to add 2.0 to new_var, an error is output. 

In [17]:
new_var + 2.0

TypeError: can only concatenate str (not "float") to str

Since new_var is defined as a string, '3.14' is not treated as a numerical value, but instead as a string of text. As a result, 5.14 is not produced as the output when 2.0 is added to new_var. A string cannot be numerically added to an int or float value. 

## Converting between data types
At times, programs will only take in certain data types as input. For example, if a program requires you to enter a variable as an integer, but the value you are trying to enter is a float (ex. 2.3456), you will need to convert that variable to an integer before entering it into the program so that it will run properly. 

To convert between data types, simply type

```python3
{new_data_type}({variable})
```

For example, to convert our float variable *var* to an int variable, we can type

In [18]:
int(var)

3

As a result *var* is interpreted as the integer (int) value 3. *NOTE* that this does not *permanently* redefine *var* as an integer value. *var* is still assigned to the float 3.14. 